### ISSUE TITLE
Lagging issues

### ISSUE THREAD

**Issue by desmandias** [TerraForged/tracker#52]
**PLEASE PROVIDE DETAILED INFORMATION, LOGS, CRASH-REPORTS.**  
[Search for existing issues before opening a new one.](https://github.com/TerraForged/tracker/issues?q=)  
_Note: we are not accepting feature requests/questions. Just bugs pls :]_

Hi! I have a custom modpack that I run both singleplayer, and on a private server. As you'd expect, this includes Terraforged but Terraforged also seems to have a huge amount of lag, particularly with shutting the world down and respawning. I know it is, in fact, related to terraforged, as removing it cleared all of the lag that was coming from both respawning and shutting down the world, but unfortunately this messed up the biome ids of my world so I cannot permantly remove it. I found a solution temporarily that "broke" the world gen for terraforge and restored it to standard vanilla in new chunks and this also pretty much cut down on lag, except for shutting down the world.

Is there any particular config settings that would help minimize or remove this? My server has eight gigs of ram at its disposal and I feel it should be able to generate this stuff just fine. I tried increases how fast or slowing down the world gen with the performance config but it didn't seem to have an effect.

**Comment by Won-Ton** [comment]
RE: Respawns:
Someone else reported lag when respawning and the profiler data they recorded showed that each time the spawn chunks were requested it was running the generation code rather than simply loading the already generated chunks. This is not normal working behaviour and is happening externally of TerraForged (we have nothing to do with loading/saving chunks).

The only time that^ that would happen is when the biome data for those chunks contain invalid biome ids, which I guess could have been caused by adding/removing mods that add biomes (such as TerraFroged). I'm guessing Minecraft doesn't fixed the invalid biomes stored on disk and just reruns the gen code every time it tries to load those chunk(s), which could indeed be a cause of lag because it's happening on the game thread rather than the world-generation thread(s).

What I suggested to that person was to move their spawn to newly generated chunks, since those should generate with valid biome ids. It could also help to move spawn to a location that has easy to generate terrain, like a flat low-altitude location with only vanilla biomes around. They never got back to me on whether they had success or not.

RE: Server hardware:
The amount of ram you need is more for keeping chunks around players loaded, and won't speed up a struggling CPU. TerraForged needs decent, fast CPU cores and the more of them the better. A lot of hosted minecraft server providers give you a single shared cpu core and only offer the option of increasing the ram capacity which will not be sufficient for a good experience.

RE: Slow to shutdown:
Third party mods could be keeping chunks loaded and somehow cause them to be slow to unload. If you have a poor CPU in your server there could be a big backlog of chunks being generated which have to be flushed to disk before the server can stop. As mentioned above, loading/saving chunks happens externally of TerraForged. Minecraft's world loading/saving system is notoriously problematic. Can't really help with that :/

RE: Modpacks:
It's worth keeping in mind that other mods can have a massive effect on worldgen performance. TerraForged on its own can perform as fast and even faster than vanilla world-gen when on decent hardware because we scale better on more cores and highly optimize our own code. It only takes one mod that doesn't take such measures to bring that performance way down :(

----

TLDR; I'm not dimissing your issues as infactual, but what you're experiencing is definitely not the normal behaviour/performance and there could be a whole range of factors at play. Hopefully my ramblings will have given you avenues to explore.

**Comment by desmandias** [comment]
it did!! thanks for a response; as for the moving spawn location, i've tried this on numerous worlds , both serverside and singleplayer to the same result as above. The one we're playing on server-wise is the third world thus far, before I could pinpoint the lag. My PC certainly isn't the best, and I'm unsure about the CPU of my serverhosting service, but my friends who do have much better computers seem to be getting the same issue when running singleplayer. Though I admit that my tests probably weren't the most thorough. 

Is there any known incompatiblities that could cause an effect like this, such as particular strucutre mods? For biomes, all I've been using is biomes o' plenty in tangent with terraforge, along with betterendforge, but I believe both of these should be compatible.

**Comment by Won-Ton** [comment]
I don't know of any mod (and can't think of anything a mod would need to do) that would cause the issue I described. That said, I've seen some mods do some pretty horrendous things in other areas of the game so all bets are off! XD

Biomes o'plenty is one of the main mods we test against for compatibility, so no issues there. Betterendforge has had some unrelated compatibility issues with TF but they've been pretty proactive about patching them, + none that affect overworld to my knowledge.

Your server will almost certainly be hosted on a single thread in that case. MC hosts are usually pretty opaque about the hardware they allot to you, and when they do give you something out of the ordinary, like 'x number of guaranteed cores', they'll be very vocal about it.
With TF, we target Mojang's minimum spec requirement of 4-core cpus. Anything lower is going to start to be a bit of lag-fest unfortunately, particularly with multiple players at once.

**Comment by desmandias** [comment]
yikes, that is unfortunate. a real shame too, these worlds were generating really pretty; thanks for your help
